src/main/java/com/puppycrawl/tools/checkstyle/site/ExampleMacro.java

```diff
@@ -0,0 +1,133 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2023 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.site;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import org.apache.maven.doxia.macro.AbstractMacro;
+import org.apache.maven.doxia.macro.Macro;
+import org.apache.maven.doxia.macro.MacroExecutionException;
+import org.apache.maven.doxia.macro.MacroRequest;
+import org.apache.maven.doxia.sink.Sink;
+import org.apache.maven.doxia.sink.impl.SinkEventAttributeSet;
+import org.codehaus.plexus.component.annotations.Component;
+
+/**
+ * A macro that inserts a snippet of code or configuration from a file.
+ */
+@Component(role = Macro.class, hint = "example")
+public class ExampleMacro extends AbstractMacro {
+
+    /** Starting delimiter for config snippets. */
+    private static final String XML_CONFIG_START = "/*xml";
+
+    /** Ending delimiter for config snippets. */
+    private static final String XML_CONFIG_END = "*/";
+
+    /** Starting delimiter for code snippets. */
+    private static final String CODE_SNIPPET_START = "// xdoc section -- start";
+
+    /** Ending delimiter for code snippets. */
+    private static final String CODE_SNIPPET_END = "// xdoc section -- end";
+
+    /** Newline character. */
+    private static final String NEWLINE = "\n";
+
+    /** Eight whitespace characters. All example source tags are indented 8 spaces. */
+    private static final String INDENTATION = "        ";
+
+    @Override
+    public void execute(Sink sink, MacroRequest request) throws MacroExecutionException {
+        final String path = (String) request.getParameter("path");
+        final String type = (String) request.getParameter("type");
+        final List<String> lines = readFile("src/xdocs-examples/" + path);
```

Comment:
Can you provide the link to the base macro doxia created I pointed you to?

One of my concerns is we will be reading the same file twice. Once for config, and second for code. We can cache the last loaded file, and just re-use it instead of reloading things from scratch.

Reply:
If confirmed .... Any optimizations for performance better to move to separate issue to be done separately. We need this PR be merged sooner to unblock bunch of other activities.

Reply:
This is the SnippetMacro: https://github.com/apache/maven-doxia/blob/24ab8ba193ec44007333775576277f9aa9ea165f/doxia-core/src/main/java/org/apache/maven/doxia/macro/snippet/SnippetMacro.java#L44

@rnveach I added a cache with a simple if-statement.

Reply:
Thanks, what I wanted to confirm was that doxia's own macro has cache built in too.
https://github.com/apache/maven-doxia/blob/24ab8ba193ec44007333775576277f9aa9ea165f/doxia-core/src/main/java/org/apache/maven/doxia/macro/snippet/SnippetMacro.java#L203